packages/app/src/composables/useCloudSpecData.ts

```diff
@@ -0,0 +1,119 @@
+ * @param allSpecs All project spec entries
+ * @returns Trigger functions
+ */
+export function useCloudSpecData (
+  isProjectDisconnected: Ref<boolean>,
+  isOffline: Ref<boolean>,
+  projectId: string | null | undefined,
+  mostRecentUpdate: Ref<string | null>,
+  displayedSpecs: Ref<(SpecsListFragment | undefined)[]>,
+  allSpecs: (SpecsListFragment | undefined)[],
+) {
+  const refetchMutation = useMutation(CloudData_RefetchDocument)
+
+  const isCloudSpecOlderThan = (item: NonNullCloudSpec, comparisonDttm: string | null) => {
+    if (item.data?.__typename !== 'CloudProjectSpec' && item.data?.__typename !== 'CloudProjectSpecNotFound') {
+      return false
+    }
+
+    if (!item.data.retrievedAt || !comparisonDttm) {
+      return false
+    }
+
+    return new Date(comparisonDttm).getTime() > new Date(item.data.retrievedAt).getTime()
+  }
+
+  const shouldRefetch = (item: NonNullCloudSpec) => {
+    if (isOffline.value) {
+      // Offline, no need to refetch
+      return false
+    }
+
+    if (item.fetchingStatus === 'NOT_FETCHED' || item.fetchingStatus === undefined) {
+      // Not yet fetched
+      return true
+    }
+
+    if (isCloudSpecOlderThan(item, mostRecentUpdate.value)) {
+      // outdated
+      return true
+    }
+
+    // nothing new, no need to refetch
+    return false
+  }
+
+  /**
+   * Refetch any displayed RemoteFetchable entries that are older than the latest cloudProject update
+   * or that have not yet been fetched
+   */
+  const fetchDisplayedCloudData = async () => {
+    const cloudSpecIdsToRefetch = displayedSpecs.value
+    .map((spec) => spec?.cloudSpec)
+    .filter((cloudSpec): cloudSpec is NonNullCloudSpec => Boolean(cloudSpec && shouldRefetch(cloudSpec)))
+    .map((cloudSpec) => cloudSpec.id)
+    ?? []
+
+    if (!isProjectDisconnected.value && !refetchMutation.fetching.value && cloudSpecIdsToRefetch.length > 0) {
+      await refetchMutation.executeMutation({ ids: cloudSpecIdsToRefetch })
+    }
+  }
+
+  /**
+   * Refetch any cloudSpec entries that are in an Error state
+   */
+  const refetchFailedCloudData = async () => {
+    const latestRunsIds = allSpecs
+    .map((s) => s?.cloudSpec)
+    .filter((cloudSpec): cloudSpec is NonNullCloudSpec => Boolean(cloudSpec?.fetchingStatus === 'ERRORED'))
+    .map((cloudSpec) => cloudSpec.id) ?? []
+
+    await refetchMutation.executeMutation({ ids: [...latestRunsIds] })
+  }
+
+  const displayedSpecIds = computed(() => displayedSpecs.value.map((v) => v?.cloudSpec?.id).join('|'))
+  const debouncedDisplayedSpecIds = useDebounce(displayedSpecIds, 200)
+
+  /*
+  Automatically trigger refresh & purge on the following:
+  - Set of displayed specs changes (scroll thru virtualized list)
+  - Network connectivity state changes
+  - Project connectivity state changes
+  - Latest update timestamp for cloud project changes
+  */
+  watch(
+    [debouncedDisplayedSpecIds, isOffline, isProjectDisconnected, mostRecentUpdate],
```

Comment:
This is interesting, I never thought about watching a debounced value. The way I've done this previously is to do something [this like](https://github.com/lmiller1990/vuejs-composition-api-v3/blob/master/src/components/PostWriter.vue#L44-L52) where you debounce the watch callback.

This is interesting too - if it works, it works! I like this composable - pretty nice way to do a kind of "if any of these things change, re-run the fetch".